core/store/src/trie/shard_tries.rs

```diff
@@ -458,21 +461,49 @@ impl ShardTries {
                 let _timer = metrics::MAKE_STATE_SNAPSHOT_ELAPSED.start_timer();
                 // `write()` lock is held for the whole duration of this function.
                 // Accessing the snapshot in other parts of the system will fail.
-                let mut state_snapshot_lock = self.0.state_snapshot.write().unwrap();
+                let mut state_snapshot_lock = self.0.state_snapshot.write().map_err(|_| {
+                    anyhow::Error::msg("error accessing write lock of state_snapshot")
+                })?;
+                let db_snapshot_hash = self.get_state_snapshot_hash();
+
                 if let Some(state_snapshot) = &*state_snapshot_lock {
-                    if &state_snapshot.prev_block_hash == prev_block_hash {
+                    // only return Ok() when the hash stored in STATE_SNAPSHOT_KEY and in state_snapshot_lock and prev_block_hash are the same
+                    if db_snapshot_hash.is_ok()
+                        && db_snapshot_hash.unwrap() == *prev_block_hash
+                        && state_snapshot.prev_block_hash == *prev_block_hash
+                    {
                         tracing::warn!(target: "state_snapshot", ?prev_block_hash, "Requested a state snapshot but that is already available");
                         return Ok(());
                     } else {
-                        let prev_block_hash = state_snapshot.prev_block_hash;
                         // Drop Store before deleting the underlying data.
                         *state_snapshot_lock = None;
-                        self.delete_state_snapshot(
-                            &prev_block_hash,
-                            home_dir,
-                            hot_store_path,
-                            state_snapshot_subdir,
-                        );
+
+                        // This will delete all existing snapshots from file system. If failed, will retry until success
+                        let mut delete_state_snapshots_from_file_system = false;
+                        while !delete_state_snapshots_from_file_system {
```

Comment:
Maybe this should be limited to retrying a certain number of times.
If the operation failed once, it's likely to keep failing. And if it fails forever, the node is in a worse state than before.

Reply:
makes sense.